fuse_viz/src/relative_pose_2d_stamped_constraint_visual.cpp

```diff
@@ -137,6 +140,61 @@ void RelativePose2DStampedConstraintVisual::setConstraint(
   error_line_->addPoint(absolute_position_ogre);
   error_line_->addPoint(toOgre(pose2.getOrigin()));
 
+  // Set error line color brightness based on the loss function impact on the constraint cost:
+  auto loss_function = constraint.lossFunction();
+  if (loss_function)
+  {
+    // Evaluate cost function without loss:
+    const double position1[] = { pose1.getOrigin().getX(), pose1.getOrigin().getY() };
+    const double yaw1[] = { tf2::getYaw(pose1.getRotation()) };
+    const double position2[] = { pose2.getOrigin().getX(), pose2.getOrigin().getY() };
+    const double yaw2[] = { tf2::getYaw(pose2.getRotation()) };
+
+    const double* parameters[] = { position1, yaw1, position2, yaw2 };
+
+    auto cost_function = constraint.costFunction();
+
+    fuse_core::VectorXd residuals(cost_function->num_residuals());
+
+    cost_function->Evaluate(parameters, residuals.data(), nullptr);
+    delete cost_function;
+
+    // The cost without the loss would be:
+    //
+    // cost = 0.5 * squared_norm
+    //
+    // See https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L159
+    const auto squared_norm = residuals.squaredNorm();
+
+    // Evaluate the loss as in:
+    // https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L164
+    //
+    // The cost with the loss would be:
+    //
+    // loss_cost = 0.5 * rho[0]
+    //
+    // See https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L165
+    double rho[3];
+    loss_function->Evaluate(squared_norm, rho);
+    delete loss_function;
+
+    // Interpolate between the constraint's absolute position and its second variable position by the quotient between
+    // the cost with and without loss:
+    //
+    //              loss_cost      0.5 * rho[0]         rho[0]
+    // loss_scale = --------- = ------------------ = ------------
+    //                cost      0.5 * squared_norm   squared_norm
+    //
+    // Remember that in principle `rho[0] <= squared_norm`, with `rho[0] == squared_norm` for the inlier region, and
+    // `rho[0] < squared_norm` for the outlier region:
+    loss_scale_ = squared_norm == 0.0 ? 0.0 : rho[0] / squared_norm;
```

Comment:
I'm trying to decide what the scale factor should be in the case where `squared_norm == 0`. If `rho[0]` is also zero, then it seems like the loss function did not modify the cost, and the scale should be 1.0. If `rho[0]` is not equal to zero, then something bad happened? Not sure what the color should be in that case.

Reply:
Good catch. The scale should be `1.0` when the `squared_norm == 0`.

If `rho[0] != 0` when `squared_norm == 0`, something bad happened, but maybe we can rely on the unit tests in `fuse_loss` to confirm that `rho[0] == 0` when `squared_norm == 0`? :thinking: 

Indeed, if `squared_norm == 0` but `rho[0] != 0`, that violates the fact that `rho[0] <= squared_norm`, which I think should always be satisfied.

There are also other case that aren't considered, like negatives values in `rho[0]`, which will also be bad. Maybe for all this cases I could add a messages like this: `ROS_WARN_STREAM_THROTTLE(10.0, "Detected invalid loss cost " << rho[0]);` :thinking:

Reply:
This is visualization code. I'm not particularly concerned about issuing warnings for invalid loss function implementations here. If we were going to do that, I feel like that should get handled in the loss function base class, or Ceres, or something that gets executed all the time.

I'll leave it to you to decide if/how you want to visualize these cases. Ignore them, set the color to black, whatever you want.

Reply:
I've set the scale to `1.0` if `sqared_norm == 0` and warn if `rho[0] > squared_norm`, and in that case clamp `rho[0]` to `squared_norm`.

That should be fine for now. I can revisit later if I find a better way. :smiley: 

I believe Ceres doesn't check any of that for efficiency reasons, or maybe because in some cases it could still be allowed that `rho[0] > squared_norm` :thinking: